packages/lodestar/src/network/encoder.ts

```diff
@@ -0,0 +1,129 @@
+import {IBeaconConfig} from "@chainsafe/lodestar-config";
+import {ERR_INVALID_REQ, MAX_CHUNK_SIZE, Method, ReqRespEncoding} from "../constants";
+import {RequestBody, ResponseBody} from "@chainsafe/lodestar-types";
+import {IReqRespEncoder, SnappyEncoder, SszEncoder} from "./encoders";
+import {getRequestMethodSSZType, getResponseMethodSSZType} from "./util";
+import * as varint from "varint";
+import {ResponseChunk} from "./interface";
+// eslint-disable-next-line import/no-extraneous-dependencies
+import BufferList from "bl";
+
+export class ReqRespEncoder {
+
+  public readonly encoding: ReqRespEncoding;
+
+  private readonly config: IBeaconConfig;
+
+  constructor(config: IBeaconConfig, encoding: ReqRespEncoding) {
+    this.config = config;
+    this.encoding = encoding;
+  }
+
+  public encodeRequest(method: Method, data: RequestBody): Buffer {
+    const type = getRequestMethodSSZType(this.config, method);
+    const encoders = this.getEncoders(this.encoding);
+    const encodedPayload =  encoders.reduce((result: unknown, encoder) => {
+      return encoder.encode(type, result);
+    }, data) as Buffer;
+    return this.writeLengthPrefixed(encodedPayload);
+  }
+  
+  public decodeRequest(method: Method, data: Buffer): RequestBody {
+    data = this.readLengthPrefixed(data);
+    const type = getRequestMethodSSZType(this.config, method);
+    //decoding is done backwards
+    const encoders = this.getEncoders(this.encoding).reverse();
+    return encoders.reduce((result: unknown, encoder) => {
+      return encoder.decode(type, result);
+    }, data) as RequestBody;
+  }
+
+  public encodeResponse(
+    method: Method
+  ): ((source: AsyncIterable<ResponseChunk>) => AsyncGenerator<Buffer>) {
+    const encoders = this.getEncoders(this.encoding);
+    const type = getResponseMethodSSZType(this.config, method);
+    const writeLP = this.writeLengthPrefixed;
+    const writeStatus = this.writeStatus;
+    return (source: AsyncIterable<ResponseChunk>) => {
+      return (async function * () {
+        for await (const chunk of source) {
+          let encodedPayload: Buffer = Buffer.alloc(0);
+          if(chunk && (chunk.output !== null && chunk.output !== undefined)){
+            encodedPayload = encoders.reduce((result: unknown, encoder) => {
+              return encoder.encode(type, result);
+            }, chunk.output) as Buffer;
+          }
+          let status = 0;
+          if(chunk.err) {
+            status = chunk.err.message === ERR_INVALID_REQ ? 1 : 2;
+            //TODO: stop yielding
+          }
+          yield writeStatus(writeLP(encodedPayload), status);
+        }
+      })();
+    };
+  }
+
+  public decodeResponse(method: Method): ((source: AsyncIterable<Buffer|BufferList>) => AsyncGenerator<ResponseBody>) {
```

Comment:
this method may throw error at one of the below calls. Probably we want to handle try/catch at the caller side (`sendRequest`) to reject the promise?

Reply:
Not really. according to spec we need to reset stream on responder side (caller won't see exception). I was hopping that exception in async iterator will reset stream. Not sure what else we can so. If we catch exception as responder, we will process previous chunks.